File: src/ui/splash.ts
// Splash banner shown at REPL startup. Renders a block-letter "SOV" mark
// (a nod to the Sovereign AI logo, shortened to match the `sov` binary)
// next to a small info card with version, provider/auth, model, and
// bundle path. Below it: a tips line and a one-line dim footer with
// the operational details that the old multi-line banner carried
// (permissions, tools, cache, session).

import chalk from 'chalk';
import { boxify, visibleWidth } from './box.js';
import { theme } from './theme.js';

const PKG_VERSION = '0.0.1';

// "ANSI Shadow" figlet font — same letterforms throughout. Each row is
// the same width; the renderer pads short lines via padBlock.
const LOGO_LINES = [
  '  ███████╗ ██████╗ ██╗   ██╗ ',
  '  ██╔════╝██╔═══██╗██║   ██║ ',
  '  ███████╗██║   ██║██║   ██║ ',
  '  ╚════██║██║   ██║╚██╗ ██╔╝ ',
  '  ███████║╚██████╔╝ ╚████╔╝  ',
  '  ╚══════╝ ╚═════╝   ╚═══╝   ',
];

// Cyan→blue gradient that visually echoes the logo's lightning-slash
// shading. One color per logo row, top to bottom.
const LOGO_GRADIENT = [
  chalk.cyanBright,
  chalk.cyan,
  chalk.blueBright,
  chalk.blue,
  chalk.blue,
  chalk.blueBright,
];

export type SplashInfo = {
  providerLabel: string;
  authLabel: string;
  model: string;
  /** Bundle path, or null in generic-agent mode. */
  bundlePath: string | null;
  permissionMode: string;
  permissionModeNote?: string;
  toolCount: number;
  cacheOn: boolean;
  sessionLabel: string;
  exitHint: string;
};

function renderCard(info: SplashInfo, maxWidth: number): string[] {
  const t = theme.tokens;
  const title = `${t.accent('>_')} ${t.textBold('Sovereign AI')} ${t.textMuted(`(v${PKG_VERSION})`)}`;
  const auth = `${info.providerLabel} ${t.textMuted('|')} ${info.authLabel}`;
  const model = `${info.model} ${t.textMuted('(/model to change)')}`;
  const bundle = t.textMuted(
    info.bundlePath ? abbreviatePath(info.bundlePath, maxWidth) : 'no bundle',
  );
  return [title, auth, model, bundle];
}

/** Shorten a long bundle path by collapsing leading segments to "…/".
 *  Keeps the last 1–2 path segments intact (the meaningful part to a
 *  user). Returns the path unchanged when it already fits. */
function abbreviatePath(path: string, maxWidth: number): string {
  if (path.length <= maxWidth) return path;
  const segments = path.split('/').filter((s) => s.length > 0);
  if (segments.length === 0) return path;
  // Try keeping more and more trailing segments; pick the longest tail
  // that fits with the "…/" prefix.
  for (let keep = Math.min(segments.length, 3); keep >= 1; keep--) {
    const tail = segments.slice(-keep).join('/');
    const candidate = `…/${tail}`;
    if (candidate.length <= maxWidth) return candidate;
  }
  // Even one segment is too long — hard-truncate.
  const last = segments[segments.length - 1] ?? path;
  return `…/${last.slice(0, Math.max(1, maxWidth - 2))}`;
}

function padBlock(lines: string[], targetHeight: number, width: number): string[] {
  const pad = ' '.repeat(width);
  const out = [...lines];
  while (out.length < targetHeight) out.push(pad);
  return out;
}

function padRight(line: string, width: number): string {
  const fill = Math.max(0, width - visibleWidth(line));
  return `${line}${' '.repeat(fill)}`;
}

/** Minimum spare-room margin between the splash content and the
 *  terminal's right edge. Without it the box-drawing characters can
 *  visually crowd the edge even when they don't strictly wrap. */
const SAFETY_MARGIN = 2;
/** Gap between the logo and the card in side-by-side layout. */
const GUTTER = 2;
/** Minimum card-content width that still looks reasonable. Below
 *  this, side-by-side isn't even attempted; we stack instead. */
const MIN_CARD_BUDGET = 30;

export function renderSplash(info: SplashInfo, terminalCols?: number): string {
  const cols = terminalCols ?? process.stdout.columns ?? 80;
  const logoWidth = LOGO_LINES[0]?.length ?? 0;
  const t = theme.tokens;

  // Layout selection — measure first, decide second.
  //
  // The original heuristic ("side-by-side if budget for the card-content
  // line is >= 30") let the layout sneak through at terminal widths
  // where the rendered row was right at the edge. A font whose box-
  // drawing characters are even slightly wider than one cell (a common
  // condition with fallback fonts on box-drawing glyphs) then pushed
  // the row past the right edge and the terminal wrapped each logo row
  // mid-glyph, fragmenting the ASCII art into apparent garbage.
  //
  // Now: build the card at the side-by-side budget, then ask whether
  // logo + gutter + actual cardWidth + safety margin fits inside cols.
  // If not, fall through to stacked. This is correct by construction
  // rather than threshold-by-trial.
  const sideBySideBudget = cols - logoWidth - SAFETY_MARGIN - 2 /* gutter */ - 4 /* card padding */;
  const stackedBudget = cols - SAFETY_MARGIN - 4 /* card padding */;
  // Build the side-by-side candidate first to measure its real width.
  const sideBySideCard =
    sideBySideBudget >= MIN_CARD_BUDGET
      ? boxify(renderCard(info, sideBySideBudget), { padding: 2 })
      : null;
  const sideBySideCardWidth = sideBySideCard ? Math.max(...sideBySideCard.map(visibleWidth)) : 0;
  const sideBySideRowWidth = logoWidth + GUTTER + sideBySideCardWidth;
  // Use side-by-side only when the row fits comfortably. SAFETY_MARGIN
  // protects against terminals that render box-drawing chars wider
  // than one cell — without it we get the regression where the row
  // visually overflows even though the cell math says it fits.
  const useStacked = sideBySideCard === null || sideBySideRowWidth + SAFETY_MARGIN > cols;

  const cardLines = useStacked
    ? boxify(renderCard(info, stackedBudget), { padding: 2 })
    : (sideBySideCard ?? boxify(renderCard(info, stackedBudget), { padding: 2 }));
  const cardWidth = Math.max(...cardLines.map(visibleWidth));

  const tips = t.textMuted(
    'Tips: type / for slash commands · @file:path to inline files · /quit to exit',
  );
  const modeNote = info.permissionModeNote ?? '';
  const footer = t.textDim(
    `perms: ${info.permissionMode}${modeNote} · tools: ${info.toolCount} · cache: ${info.cacheOn ? 'on' : 'off'} · ${info.sessionLabel} · ${info.exitHint}`,
  );

  if (useStacked) {
    // Narrow terminal: stack the logo above the card vertically. Drop
    // the logo entirely when it's wider than the terminal — the box-
    // drawing characters fragment when wrapped.
    const coloredLogo =
      logoWidth + SAFETY_MARGIN <= cols
        ? LOGO_LINES.map((line, i) => (LOGO_GRADIENT[i] ?? chalk.cyan)(line))
        : [];
    return [
      '',
      ...coloredLogo,
      ...(coloredLogo.length > 0 ? [''] : []),
      ...cardLines,
      '',
      tips,
      footer,
      '',
    ].join('\n');
  }

  // Side-by-side layout: logo on the left, card on the right.
  const coloredLogo = LOGO_LINES.map((line, i) => {
    const tint = LOGO_GRADIENT[i] ?? chalk.cyan;
    return tint(line);
  });
  const height = Math.max(coloredLogo.length, cardLines.length);
  const left = padBlock(coloredLogo, height, logoWidth);
  // Vertically center the card against the logo.
  const cardOffset = Math.max(0, Math.floor((height - cardLines.length) / 2));
  const right: string[] = Array(height).fill('');
  for (let i = 0; i < cardLines.length; i++) {
    right[cardOffset + i] = cardLines[i] ?? '';
  }
  const rows = left.map((l, i) => `${l}  ${padRight(right[i] ?? '', cardWidth)}`);
  return ['', ...rows, '', tips, footer, ''].join('\n');
}

File: packages/tui/internal/components/splash.go
// Package components — Splash: TUI startup banner.
//
// Mirrors the REPL splash (src/ui/splash.ts) — the "ANSI Shadow" SOV
// block-letter logo with a cyan→blue vertical gradient. Rendered into
// the transcript on Init so users see the same brand cue regardless of
// which surface they're on. M11.1.

package components

import (
	"strings"

	"github.com/charmbracelet/lipgloss"
	"github.com/yevgetman/sovereign-ai-harness/packages/tui/internal/style"
	"github.com/yevgetman/sovereign-ai-harness/packages/tui/internal/theme"
)

// logoLines is the "ANSI Shadow" figlet font for "SOV". Every row is the
// same width; renderers can pad on the right when they want a fixed block.
var logoLines = []string{
	"  ███████╗ ██████╗ ██╗   ██╗ ",
	"  ██╔════╝██╔═══██╗██║   ██║ ",
	"  ███████╗██║   ██║██║   ██║ ",
	"  ╚════██║██║   ██║╚██╗ ██╔╝ ",
	"  ███████║╚██████╔╝ ╚████╔╝  ",
	"  ╚══════╝ ╚═════╝   ╚═══╝   ",
}


// SplashInfo carries the optional info-card fields rendered next to the
// logo when the terminal is wide enough. Empty fields are skipped.
type SplashInfo struct {
	Version  string
	Provider string // e.g., "anthropic"
	Auth     string // e.g., "API Key"
	Model    string // e.g., "claude-haiku-4-5-20251001"
	Cwd      string // working directory (or bundle path) — abbreviated when long
	Tips     string // tips line under the logo (e.g., "type / for slash commands…")
}

// RenderSplash returns a multi-line splash string for the given width.
// On narrow terminals the logo stacks above the info card; on wide
// terminals they sit side-by-side. The width should be the renderable
// content width (terminal columns minus any frame padding).
func RenderSplash(info SplashInfo, t theme.Theme, width int) string {
	if width <= 0 {
		width = 80
	}
	logoWidth := visualWidthMax(logoLines)
	gutter := style.S.Splash.Gutter
	safetyMargin := style.S.Splash.SafetyMargin

	// Build the info card once and measure it. M11.3 — wrap in a
	// rounded lipgloss border so the card reads as a discrete element
	// (mirrors the REPL splash + the Qwen Code reference layout).
	cardContent := buildInfoCard(info, t)
	var card []string
	if len(cardContent) > 0 {
		// ux-fixes round 3: Padding(1, 2) — 1 row top + bottom and 2
		// cols left + right — gives the info card room to breathe.
		// The pre-existing Padding(0, 1) was visually cramped against
		// the rounded border (ux1.png feedback).
		borderStyle := lipgloss.NewStyle().
			Border(style.S.Card.Border).
			BorderForeground(t.Border).
			Padding(style.S.Card.GenerousPaddingV, style.S.Card.GenerousPaddingH)
		boxed := borderStyle.Render(strings.Join(cardContent, "\n"))
		card = strings.Split(boxed, "\n")
	}
	cardWidth := visualWidthMax(card)

	useStacked := width < logoWidth+gutter+cardWidth+safetyMargin

	var rows []string
	if useStacked {
		// Stack: logo on top, card below. Drop the logo entirely when even
		// the logo can't fit (avoids fragmenting box-drawing glyphs on
		// pathologically narrow terminals).
		if logoWidth+safetyMargin <= width {
			rows = append(rows, colorize(logoLines)...)
			rows = append(rows, "")
		}
		rows = append(rows, card...)
	} else {
		// Side-by-side: pad the shorter block to the taller block's height
		// so the join sees equal-length slices. Center the card vertically
		// against the logo for a balanced look.
		colored := colorize(logoLines)
		height := max(len(colored), len(card))
		left := padBlock(colored, height, logoWidth)
		cardOffset := max(0, (height-len(card))/2)
		right := make([]string, height)
		pad := strings.Repeat(" ", cardWidth)
		for i := 0; i < height; i++ {
			cardIdx := i - cardOffset
			if cardIdx >= 0 && cardIdx < len(card) {
				line := card[cardIdx]
				fill := cardWidth - lipgloss.Width(line)
				if fill < 0 {
					fill = 0
				}
				right[i] = line + strings.Repeat(" ", fill)
			} else {
				right[i] = pad
			}
		}
		for i := 0; i < height; i++ {
			rows = append(rows, left[i]+strings.Repeat(" ", gutter)+right[i])
		}
	}

	if info.Tips != "" {
		rows = append(rows, "")
		tipsStyle := lipgloss.NewStyle().Foreground(t.Dim).Italic(true)
		rows = append(rows, tipsStyle.Render(info.Tips))
	}

	return strings.Join(rows, "\n")
}

// buildInfoCard returns the info-card lines (without an outer border —
// the splash uses a label/value layout for compactness). Empty fields
// are skipped so a fresh-session boot before status_update arrives
// renders a logo + tips line cleanly.
func buildInfoCard(info SplashInfo, t theme.Theme) []string {
	accent := lipgloss.NewStyle().Foreground(t.Primary).Bold(true)
	muted := lipgloss.NewStyle().Foreground(t.Dim)
	// ux-fixes round 3: body text inherits terminal default foreground
	// per the M11.10 rule (docs/conventions/tui-color-rendering.md). A
	// hard `Foreground(t.Foreground)` rendered dark grey on terminals
	// whose palette mapped the theme foreground to a low-contrast slot
	// (ux1.png feedback). Bold-only is the "brightest reliable" path.
	body := lipgloss.NewStyle().Bold(true)
	plain := lipgloss.NewStyle()

	title := accent.Render(">_") + " " + body.Render("Sovereign AI")
	if info.Version != "" {
		title += " " + muted.Render("("+info.Version+")")
	}
	rows := []string{title}

	if info.Provider != "" || info.Auth != "" {
		line := info.Provider
		if info.Provider != "" && info.Auth != "" {
			line += " " + muted.Render("|") + " " + info.Auth
		} else if info.Auth != "" {
			line = info.Auth
		}
		rows = append(rows, plain.Render(line))
	}

	// Model line. Skip when empty OR equal to the legacy "?" sentinel so
	// a fresh boot before the model name is known doesn't render a
	// confusing "? (/model to change)" line (ux1.png feedback).
	if info.Model != "" && info.Model != "?" {
		rows = append(rows, plain.Render(info.Model)+" "+muted.Render("(/model to change)"))
	}

	if info.Cwd != "" {
		rows = append(rows, muted.Render(info.Cwd))
	}

	return rows
}

// colorize applies the per-row gradient to the logo lines.
func colorize(lines []string) []string {
	gradient := style.S.Brand.LogoGradient
	out := make([]string, len(lines))
	for i, line := range lines {
		c := lipgloss.Color(gradient[i%len(gradient)])
		out[i] = lipgloss.NewStyle().Foreground(c).Render(line)
	}
	return out
}

// padBlock right-pads each line to width and appends empty lines until
// the block reaches the target height. Mirrors src/ui/splash.ts.
func padBlock(lines []string, height, width int) []string {
	out := make([]string, height)
	pad := strings.Repeat(" ", width)
	for i := 0; i < height; i++ {
		if i < len(lines) {
			line := lines[i]
			fill := width - lipgloss.Width(line)
			if fill < 0 {
				fill = 0
			}
			out[i] = line + strings.Repeat(" ", fill)
		} else {
			out[i] = pad
		}
	}
	return out
}

// visualWidthMax returns the widest visual width across the given lines.
// Uses lipgloss.Width which handles ANSI escape codes and wide chars.
func visualWidthMax(lines []string) int {
	w := 0
	for _, line := range lines {
		if lw := lipgloss.Width(line); lw > w {
			w = lw
		}
	}
	return w
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
